src/expressions/DynamicContext.ts

```diff
@@ -16,6 +16,8 @@ class DynamicContext {
 	public contextItemIndex: number;
 	public contextSequence: ISequence;
 	public variableBindings: { [s: string]: () => ISequence };
+	public generateIdMap: WeakMap<Node, string>;
```

Comment:
If you move this weakmap to the module level, as in, make it its own file like this:

```
const generateIdMap: WeakMap<Node,string> = new WeakMap();
export default generateIdMap;
```

and just directly talk to that in the generate-id function, circumventing the DynamicContext, you will get that stability over multiple calls. Because this is a weakMap I don't expect any memory downsides. I'd like this more than externalizing this state with a DomFacade. That would require other users of FontoXPath to do some upgrading on their side, whcih I'm trying to minimize.

Reply:
Makes sense to me! I'll give this a shot. Thank you

Reply:
I implemented the function using a module, but the `WeakMap` is not working as I expected with nodes. It doesn't seem to be recognizing previous nodes. I'm trying to track this down, but I would have expected nodes to be equal across subsequent calls. See the failing test here https://github.com/FontoXML/fontoxpath/actions/runs/9475980881/job/26108159273?pr=641#step:6:100

Reply:
NodePointers. Forgot about them.

In XQuery (Update Facility) we are sometimes cloning nodes, a lot. When you wrap a large structure (like `let $ele := <ele/> let $wrap := <wrap>{$ele}</wrap> return $wrap`) the `<ele/>` in `$ele` and the one in the `$wrap` are not equal (`$ele is $wrap/ele` is false). [playground](https://xpath.playground.fontoxml.com/?state=H4sIAEQ9aWYAA21Ry2rDMBD8lcUU0kKw6TU4ht5TaEsLKfii2OtYVC%2BkVWoT8u9dyQ05tCDQrDQ7O9KcC217LDaP6%2BIkvBQHhaHYFOdLsS4mrRjWvDWtAahJksJmcoLG0ikxH72Npi8Ha8gyqeysrquFlPkhai383LyPMgAvAQqFN9IcgaxVMFgP%2B%2BfdGvYvLAnC9LB%2Fjejnsq6uvb%2BDXchowSD77Qp7Savm00bohIFUAZ64l8akbw3QiDxvoGTJ%2FemehHb81JuC6DoMAbT1CNdbGLzVbFujiSAXSbIOvDyO%2F%2Bs69Fooab5WzZOxzPdwiESLHVZWSAFmHhlGwSUjD4SBIIZkm318vO1uyhmlp9dVSiFlkj%2BIY2EluEOFsNlCzXvFrHz27YXLhwk058S51FUuWuORojdw35qWcjfHkjsqLlrzwBM6jhMnus4pLj%2BU9WCiIwIAAA%3D%3D)

Figuring out how to circumvent this. The `nodeCompare` function compares node identity, so that's close. We might have to memoize the `createPointer` function in DomFacade.ts here. That is the only place we make NodePointers. This way it can work over XQuery-made nodes as well.

Reply:
This seems to work for the tests I wrote. There are probably situations where it won't work:

```
import { NodePointer, TinyNode } from '../../domClone/Pointer';
import { ConcreteNode } from 'src/domFacade/ConcreteNode';

const generateIdMap: WeakMap<TinyNode | ConcreteNode, string> = new WeakMap();
let generateIdCounter = 0;

function generateId(ptr: NodePointer): string {
  const node = ptr.node;
	if (!generateIdMap.has(node)) {
		generateIdMap.set(node, `id${++generateIdCounter}`);
	}
	return generateIdMap.get(node);
}

export default generateId;
```